internal/websocket/e2e_test.go

```diff
@@ -23,6 +25,8 @@ import (
 )
 
 func TestEndToEnd(t *testing.T) {
+	var finished bool
```

Comment:
Consider using a test suite, as implemented by https://pkg.go.dev/github.com/stretchr/testify/suite.  This will clean up this test code considerably, as you can move a lot of setup code into the lifecycle of the test rather than the test method.